src/go/k8s/internal/controller/redpanda/cluster_controller.go

```diff
@@ -1023,35 +981,69 @@ func (r *ClusterReconciler) doDecommissionGhostBrokers(c context.Context, vClust
 	if err != nil {
 		return fmt.Errorf("unable to create admin client: %w", err)
 	}
-	adminBrokers, err := adminClient.Brokers(ctx)
+
+	health, err := adminClient.GetHealthOverview(ctx)
+	if err != nil {
+		return fmt.Errorf("feating cluster heatlh overview: %w", err)
+	}
+
+	// Until cluster does not report that its healthy do not decommission
+	if !health.IsHealthy {
+		err = errors.New("cluster is not healthy") //nolint:goerr113,err113 // That error will not be handled by reconciler function
+		log.V(logger.DebugLevel).Error(err, "stopping decommission verification",
+			"requested-replicas", vCluster.Spec.Replicas,
+			"nodes-registered", health.AllNodes,
+			"nodes-down", health.NodesDown)
+		return err
+	}
+
+	pods, err := r.podList(ctx, vCluster)
 	if err != nil {
-		return fmt.Errorf("unable to fetch brokers: %w", err)
+		return fmt.Errorf("unable to fetch PodList: %w", err)
 	}
 
+	// Create map of Redpanda IDs from Pod resource
 	actualBrokerIDs := make(map[int]any, len(pods.Items))
 	for i := range pods.Items {
 		pod := &pods.Items[i]
 		if pod.Annotations == nil {
 			return fmt.Errorf("requeuing: %w", &missingBrokerIDError{})
 		}
+
 		nodeIDStrAnnotation, annotationExist := pod.Annotations[resources.PodAnnotationNodeIDKey]
 		if !annotationExist {
 			return fmt.Errorf("requeuing: %w", &missingBrokerIDError{})
 		}
-		id, err := strconv.Atoi(nodeIDStrAnnotation)
+
+		var id int
+		id, err = strconv.Atoi(nodeIDStrAnnotation)
 		if err != nil {
 			return fmt.Errorf("pod %s has an invalid broker-id annotation: %q: %w", pod.Name, nodeIDStrAnnotation, err)
 		}
+
 		actualBrokerIDs[id] = nil
 	}
 
-	// if the admin API shows brokers that are not assigned to existing pods, decommission them
-	for i := range adminBrokers {
-		broker := adminBrokers[i]
-		if _, ok := actualBrokerIDs[broker.NodeID]; ok {
-			continue
+	// Reduce brokers to the list of brokers that are not visible at the K8S level
+	bl, err := adminClient.Brokers(ctx)
+	if err != nil {
+		return fmt.Errorf("retrieving broker list: %w", err)
+	}
+
+	log.V(logger.DebugLevel).Info("removes Redpanda which reported its ID (in Pod annotation) against Redpanda broker list",
```

Comment:
let's roll this into the log line on 1042 so we get all relevant debugging information within a single log line?

If anything goes wrong, we'll want to see the list of brokers from K8s perspective, the list of brokers from redpanda's perspective, and the list of brokers that are considered to be ghosts.

Reply:
Agree